Updated cookie secrets management

wtoorop opened the pull request:
This PR is in response to issue #196 (and thus also of interest to @anandb-ripencc  )
- The cookie secret file's default location can be set with `--with-cookiesecretsfile=path` with `configure`
- The default cookie secrets file is changed to `{dbdir}/cookiesecrets.txt`
- When no cookie secrets file location is set with the `cookie-secret-file` option in the config file, and the new default location does not exist, cookie secrets will be read from the previous default location `{configdir}/nsd_cookiesecrets.txt`
- A staging cookie secret can also be provided in the config file with the `cookie-staging-secret` option.
- Explicitly configured cookie secrets in config file (with the `cookie-secret` and `cookie-staging-secret` option) take precedence over the ones from the cookie secret file
- All values of DNS cookie related settings from config file (`answer-cookie`, `cookie-secret`, `cookie-staging-secret` and `cookie-secret-file`) will be reevaluated and effectuated after `nsd-control reconfig`

wtoorop committed:
--with-cookiesecretsfile=path for configure

Partly addressing issue #196

wtoorop committed:
Configurable staing cookie

wtoorop committed:
Allow cookie settings to be reconfigurable

wtoorop committed:
Automatic migration of cookie secrets

from the old to the new default location

wtoorop committed:
Updated documentation on cookie server secrets

wtoorop committed:
Tell were cookies came from

even when they came from the old default location
+some small code cleanup and reduced code repetition

wtoorop commented:
It's ready for review (and merge if accepted). One thing to consider/discuss is that it has one thing that is not backwards compatible, which is that cookie secrets from config file now are preferred over cookies from the cookie secrets file. Before this was the other way around. But, this feels more logical to me as zones in the configuration file are also preferred over zones in the zone list file. I believe there will be a negligible number of people that have cookie secrets in both the config file and in the cookie secrets file and would depend on the earlier preference to pick the secrets file.

anandb-ripencc commented:
Looks reasonable to me. I agree that despite the backwards-incompatible change, it is unlikely to affect anyone, because no-one would define cookies in both places.

wcawijngaards commented on difffile.c:
```diff
@@ -1988,53 +1969,13 @@ task_process_del_key(struct nsd* nsd, struct task_list_d* task)
+	
+	nsd->do_answer_cookie = (int) task->newserial;
+	nsd->cookie_count = (size_t) task->yesno;
+	memmove(nsd->cookie_secrets, task->zname, sizeof(nsd->cookie_secrets));
```
The function `task_process_add_cookie_secret` that was deleted used `explicit_bzero`, to wipe the secrets after use. The change adds that to this function.

```suggestion
	memmove(nsd->cookie_secrets, task->zname, sizeof(nsd->cookie_secrets));
	explicit_bzero(task->zname, sizeof(nsd->cookie_secrets));
```

wcawijngaards commented on remote.c:
```diff
@@ -2366,65 +2430,75 @@ cookie_secret_file_dump(RES* ssl, nsd_type const* nsd) {
+		memcpy( nsd->cookie_secrets, backup_cookie_secrets
+		      , sizeof(cookie_secrets_type));
+		nsd->cookie_count = backup_cookie_count;
+	}
```
The `backup_cookie_secrets` variable can be erased after the function is done.

```suggestion
	}
	explicit_bzero(backup_cookie_secrets, sizeof(cookie_secrets_type));
```

wcawijngaards commented on remote.c:
```diff
@@ -2366,65 +2430,75 @@ cookie_secret_file_dump(RES* ssl, nsd_type const* nsd) {
+		memcpy( nsd->cookie_secrets, backup_cookie_secrets
+		      , sizeof(cookie_secrets_type));
+		nsd->cookie_count = backup_cookie_count;
+	}
```
The `backup_cookie_secrets` variable can be erased after the function is done.

```suggestion
	}
	explicit_bzero(backup_cookie_secrets, sizeof(cookie_secrets_type));
```

wcawijngaards commented on remote.c:
```diff
@@ -2443,27 +2517,24 @@ do_add_cookie_secret(RES* ssl, xfrd_state_type* xrfd, char* arg) {
+		memcpy( nsd->cookie_secrets, backup_cookie_secrets
+		      , sizeof(cookie_secrets_type));
+		nsd->cookie_count = backup_cookie_count;
+	}
```
The `backup_cookie_secrets` variable can be erased after the function is done.

```suggestion
	}
	explicit_bzero(backup_cookie_secrets, sizeof(cookie_secrets_type));
```

wcawijngaards commented on util.c:
```diff
@@ -1145,3 +1149,110 @@ void drop_cookie_secret(struct nsd* nsd)
+			log_msg( LOG_ERR
+			       , "error parsing cookie secret file \"%s\""
+			       , fn);
+			return 0;
```
The `cookie_secrets` and `secret` variables can be erased when the function is done.

```suggestion
			explicit_bzero(cookie_secrets, sizeof(cookie_secrets));
			explicit_bzero(secret, sizeof(secret));
			return 0;
```

wcawijngaards commented on util.c:
```diff
@@ -1145,3 +1149,110 @@ void drop_cookie_secret(struct nsd* nsd)
+			log_msg( LOG_ERR
+			       , "error parsing cookie secret file \"%s\""
+			       , fn);
+			return 0;
```
The `cookie_secrets` and `secret` variables can be erased when the function is done.

```suggestion
			explicit_bzero(cookie_secrets, sizeof(cookie_secrets));
			explicit_bzero(secret, sizeof(secret));
			return 0;
```

wcawijngaards commented on util.c:
```diff
@@ -1145,3 +1149,110 @@ void drop_cookie_secret(struct nsd* nsd)
+		region_str_replace(  nsd->region
+		                  , &nsd->cookie_secrets_filename, fn );
+	}
+	return 1;
```
The `cookie_secrets` variable can be erased when the function is done.

```suggestion
	explicit_bzero(cookie_secrets, sizeof(cookie_secrets));
	return 1;
```

wcawijngaards commented on util.c:
```diff
@@ -1145,3 +1149,110 @@ void drop_cookie_secret(struct nsd* nsd)
+			       , "error parsing cookie secret file \"%s\""
+			       , fn);
+			return 0;
+		}
```
The `secret` variable can be erased after use.

```suggestion
		explicit_bzero(secret, sizeof(secret));
		}
```

wcawijngaards reviewed (approved):
The code looks all right. I saw some places where the cookie secrets could possibly be wiped after temporary variables are done being used, maybe that is nice to have.

wtoorop committed:
Apply suggestions from code review

Thanks @wcawijngaards ! Yes, cleanup secrets on the stack after use is definitely more secure. I think you even found some occurrences that weren't covered before.

Co-authored-by: Wouter Wijngaards <wcawijngaards@users.noreply.github.com>

k0ekk0ek commented on remote.c:
```diff
@@ -2473,7 +2547,23 @@ do_print_cookie_secrets(RES* ssl, xfrd_state_type* xrfd, char* arg) {
+	case COOKIE_SECRETS_FROM_CONFIG:
+		ssl_printf(ssl, "source : configuration\n");
+		break;
+	default:
+		ssl_printf(ssl, "source : unknown\n");
+		break;
```
All possibilities are covered, so this is a panic situation(?)

wtoorop replied:
I guess yes. Would you prefer a assert here?

wtoorop replied:
Would you prefer an assert?

k0ekk0ek commented on util.c:
```diff
@@ -1145,3 +1149,116 @@ void drop_cookie_secret(struct nsd* nsd)
+		return 0;
+	}
+	else if(nsd->options->cookie_secret_file != NULL /* explicit name */
+	     ||!(f = fopen((fn = CONFIGDIR"/nsd_cookiesecrets.txt"), "r")))
```
Falling back to `nsd_cookiesecrets.txt` in this function doesn't seem right. I think it's preferred to handle that in the configuration parser (or shortly after). Maybe do something like if `cookie_secret_file` is still `NULL` after parsing, set it to the previous default? That way behavior is more deterministic. Also would avoid the `cookie_secret_file` function, which seems to return `COOKIESECRETSFILE` if the `cookie_secret_file` is `NULL`. So, basically `fn` is only `NULL` if the `cookie_secret_file` is not `NULL` but empty, in which case `NULL` is returned. This seems a very convoluted way of avoiding checking whether the member is `NULL` just after parsing and setting right then(?)

wtoorop replied:
Okay, addressed in [c743742](https://github.com/NLnetLabs/nsd/pull/378/commits/c743742a6ebdc899bb6b65b0f8ff2171662517c1)

k0ekk0ek commented on util.c:
```diff
@@ -1145,3 +1149,116 @@ void drop_cookie_secret(struct nsd* nsd)
+		}
+	}
+	fclose(f);
+	if(count && nsd->cookie_secrets_source <= COOKIE_SECRETS_FROM_FILE) {
```
An `enum` is mapped to an integer, but it's better to be explicit. The order might change and this may go unnoticed. Basically, you want the branch to be taken if the secret does not originate from the configuration (`!= COOKIE_SECRETS_FROM_CONFIG`).

wtoorop replied:
Done in [5674228](https://github.com/NLnetLabs/nsd/pull/378/commits/5674228f1cd200bbf1e5a645d55e8bbf0b90dd3d)

k0ekk0ek reviewed (approved):
I agree with the changes in general, just some small remarks regarding use of `enums` and one related to falling back to the previous value.

wtoorop commented on util.c:
```diff
@@ -1145,3 +1149,116 @@ void drop_cookie_secret(struct nsd* nsd)
+		}
+	}
+	fclose(f);
+	if(count && nsd->cookie_secrets_source <= COOKIE_SECRETS_FROM_FILE) {
```
```suggestion
	if(count && nsd->cookie_secrets_source != COOKIE_SECRETS_FROM_FILE) {
```

wtoorop committed:
Update util.c

wtoorop committed:
Merge branch 'master' into bugfix/cookiesecretsfile

wtoorop committed:
Remove order dependent comparisons with the cookie_secrets_source enum

anandb-ripencc commented:
I was reviewing the changes to the man pages and documentation, and it needs spelling and syntax fixes. Let me comment on the changes so you can fix them before merging this branch.

wtoorop commented:
> I was reviewing the changes to the man pages and documentation, and it needs spelling and syntax fixes. Let me comment on the changes so you can fix them before merging this branch.

Thanks! I'm also still addressing the last comment from @k0ekk0ek, which will simplify the code somewhat and also _only does the fallback_ to the old default location for secrets (if there is nothing configured and nothing in the new default location), _when nsd just started_.

wtoorop commented:
> I was reviewing the changes to the man pages and documentation, and it needs spelling and syntax fixes. Let me comment on the changes so you can fix them before merging this branch.

@anandb-ripencc I am done with review feedback and will merge once you're finished your spelling corrections.

anandb-ripencc commented:
Actually, just merge it. I will then submit a PR to update the documentation. Please don't make a new release before I've submitted documentation fixes.

k0ekk0ek commented:
Looks good te me @wtoorop :+1:

wtoorop commented:
I accidentally committed along changes in simdzone. I'll force push the commit without the simdzone changes, and then merge.

wtoorop committed:
Pro-actively set cookie_secret_file value

Right after config is read, so no wrapper is needed anymore to determine the value.
